src/Native/System.Security.Cryptography.Native/pal_ecdsa_import_export.cpp

```diff
@@ -0,0 +1,461 @@
+
+	if (curveType == ECCurveType::PrimeMontgomery)
+		return const_cast<EC_METHOD*>(EC_GFp_mont_method());
+
+	if (curveType == ECCurveType::Characteristic2)
+		return const_cast<EC_METHOD*>(EC_GF2m_simple_method());
+
+	return nullptr; //Edwards and others
+}
+
+extern "C" ECCurveType CryptoNative_EcKeyGetCurveType(
+	const EC_KEY* key)
+{
+	const EC_GROUP* group = EC_KEY_get0_group(key);
+	if (!group) return ECCurveType::Unspecified;
+
+	const EC_METHOD* method = EC_GROUP_method_of(group);
+	if (method == nullptr) return ECCurveType::Unspecified;
+
+	return MethodToCurveType(const_cast<EC_METHOD*>(method));
+}
+
+extern "C" int32_t CryptoNative_GetECKeyParameters(
+	const EC_KEY* key,
+	int32_t includePrivate,
+	BIGNUM** qx, int32_t* cbQx,
+	BIGNUM** qy, int32_t* cbQy,
+	BIGNUM** d,  int32_t* cbd)
+{
+	if (!key || 
+		!qx || !cbQx ||
+		!qy || !cbQy ||
+        (includePrivate && (!d  || !cbd)))
+	{
+		assert(false);
+
+		// since these parameters are 'out' parameters in managed code, ensure they are initialized
+		if (qx) *qx = nullptr; if (cbQx) *cbQx = 0;
+		if (qy) *qy = nullptr; if (cbQy) *cbQy = 0;
+		if (d)  *d  = nullptr; if (cbd) *cbd   = 0;
+
+		return 0;
+	}
+
+	// optional parameters
+    if (d)
+	    *d = nullptr;
+    
+    if (cbd)
+	    *cbd = 0;
+
+	int rc = 0;
+	BIGNUM *xBn = nullptr;
+	BIGNUM *yBn = nullptr;
+
+	ECCurveType curveType = CryptoNative_EcKeyGetCurveType(key);
+	const EC_POINT* Q = EC_KEY_get0_public_key(key);
+	const EC_GROUP* group = EC_KEY_get0_group(key); // curve
+	if (curveType == ECCurveType::Unspecified || !Q || !group) goto error;
+
+	xBn = BN_new();
+	yBn = BN_new();
+	if (!xBn || !yBn) goto error;
+
+	// extract P, A, B
+	if (curveType == ECCurveType::Characteristic2)
+	{
+		if (!EC_POINT_get_affine_coordinates_GF2m(group, Q, xBn, yBn, nullptr)) goto error;
+	}
+	else
+	{
+		if (!EC_POINT_get_affine_coordinates_GFp(group, Q, xBn, yBn, nullptr)) goto error;
+	}
+
+	// success; assign variables
+	*qx = xBn; *cbQx = BN_num_bytes(xBn);
+	*qy = yBn; *cbQy = BN_num_bytes(yBn);
+
+	if (includePrivate)
+	{
+		*d = const_cast<BIGNUM*>(EC_KEY_get0_private_key(key));
+		*cbd = BN_num_bytes(*d);
+	}
+
+	// success
+	return 1;
+
+error:
+    if (cbd) *cbd = nullptr;
+    *cbQx = *cbQy;
+    *qx = *qy;
+    if (d) *d = nullptr;
+	if (xBn) BN_free(xBn);
+	if (yBn) BN_free(yBn);
+	return rc;
+}
+
+extern "C" int32_t CryptoNative_GetECCurveParameters(
+	const EC_KEY* publicKey,
+	int32_t includePrivate,
+	ECCurveType* curveType,
+	BIGNUM** qx, int32_t* cbQx,
+	BIGNUM** qy, int32_t* cbQy,
+	BIGNUM** d, int32_t* cbd,
+	BIGNUM** P, int32_t* cbP,
+	BIGNUM** A, int32_t* cbA,
+	BIGNUM** B, int32_t* cbB,
+	BIGNUM** Gx, int32_t* cbGx,
+	BIGNUM** Gy, int32_t* cbGy,
+	BIGNUM** n, int32_t* cbn,
+	BIGNUM** h, int32_t* cbh,
+	BIGNUM** S, int32_t* cbS)
+{
+	// get the public key parameters
+	int32_t rc = CryptoNative_GetECKeyParameters(publicKey, includePrivate, qx, cbQx, qy, cbQy, d, cbd);
```

Comment:
No since CryptoNative_GetECKeyParameters exits early if one of the out parameters is null, so we want to do that before this method checks for it own 'out' parameters. I will add a comment